src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Kafka/CachedMessageHeadersHelper.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Reflection;
+using System.Reflection.Emit;
+using Datadog.Trace.DuckTyping;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Kafka
+{
+    internal static class CachedMessageHeadersHelper<TMarkerType>
+    {
+        private static readonly Func<object> _activator;
+
+        static CachedMessageHeadersHelper()
+        {
+            var headersType = typeof(TMarkerType).Assembly.GetType("Confluent.Kafka.Headers");
+
+            ConstructorInfo ctor = headersType.GetConstructor(System.Type.EmptyTypes);
+
+            DynamicMethod createHeadersMethod = new DynamicMethod(
+                $"KafkaCachedMessageHeadersHelpers",
+                headersType,
+                null,
+                typeof(DuckType).Module,
```

Comment:
With our recent robustness to assembly context's, I'm worried that creating this in the `typeof(DuckType).Module` module may not work in some edge cases. It's hard to make concrete without a repro, but the `Samples.ServiceStackRedis` program shows one example of loading the same assembly into multiple contexts and then instrumenting program runs that use each assembly. If this is something you feel should be addressed in a separate PR, we can punt this.

To resolve this, I'm thinking we can re-use some of the logic from our DuckType library found below to generate a new ModuleBuilder (which would then be supplied as the module argument). And instead of getting the generic Type `TMarkerType`, we would pass in the `topicPartition` object itself and get the assembly from the runtime type. What do you think?

https://github.com/DataDog/dd-trace-dotnet/blob/a5878ce09f443eea6165a839f09afc674ea0b357/src/Datadog.Trace/DuckTyping/DuckType.Statics.cs#L82-L96

Reply:
Just so I understand, the issue you're describing is:

* Multiple instances of the Confluent.Kafka library are loaded
* The `TMarkerType` would have a different identity for each assembly, so the static constructor would be called once for each assembly
* As we're currently creating the type in the `typeof(DuckType).Module`, we'd get an exception trying to create the second one (in the same module)

If so, then yes, I think your suggestion is worth doing. Will have to add caching using a different approach than the static constructor too I guess.

Reply:
Yes that about sums it up with one minor clarification:

> As we're currently creating the type in the typeof(DuckType).Module, we'd get an exception trying to create the second one (in the same module)
What we observed is that we wouldn't get an exception when we create the dynamic method, but when we go to execute it. The type from new instance of Confluent.Kafka would be passed into the new DynamicMethod, but because the DynamicMethod is still in the same module as the previously created one, the DynamicMethod would resolve type references to the previous (different) assembly/type and a type cast exception would occur.

I don't have a repro right now but since I've been babbling about theoretical stuff I can spend some time on Monday to come up with a repro to validate the changes.

Reply:
I think the issue is that you're trying to assign the result all back to the same `Message` object. For each assembly test, can you dynamically create a new `Message` object via reflection (in addition to the headers object) instead of re-using the original one?

Reply:
Unfortunately, I tried that in https://github.com/DataDog/dd-trace-dotnet/pull/1444/commits/c1d613706f8b578b986b1eac5d8a1fe4cc033c97 but still no failures! 🙁

Reply:
After spending several hours on this I couldn't get it to reproduce. On close inspection, I do think this scenario is different from previously failing scenarios and is safe to use with only one common Module. It's hard to remember the exact details of the previous issue but I believe it centered on the fact that we were using metadata tokens as inputs to be resolved instead of actual MethodInfo / Types that were already instantiated by the runtime. In this scenario, the DynamicMethod does not take any input arguments, it has a return type that we specify via a Type object, and the constructor we invoke has already been created as a RuntimeConstructorInfo object. So long story short, I'm like 70% sure we don't have to do anything extra here

Reply:
And I can also confirm that the message objects and header objects created in your unit test look 100% correct. Each test results in a header object that comes from the same runtime assembly as the message object.

Reply:
Just to clarify, the problem we had on the Ducktyping related to ALC wasn't caused by the `DynamicMethod`, the `DynamicMethod` always returned the right thing. 

The problem was related to the `ModuleBuilder` where we dynamically create and emit a [`delegate`](https://github.com/DataDog/dd-trace-dotnet/blob/master/src/Datadog.Trace/DuckTyping/ILHelpersExtensions.cs#L24-L51). This delegate was used later to [call the `CreateDelegate` method from the `DynamicMethod`](https://github.com/DataDog/dd-trace-dotnet/blob/master/src/Datadog.Trace/DuckTyping/DuckType.Statics.cs#L121-L125). Hence the problem if we emit the delegate for a type in one module builder we cannot reuse the same delegate if we use a dynamic method with a type in a different context. In our case was failing the casting. Same type, different context. The fix? create a new module builder so we can emit the delegate again.

Reply:
Thanks for the follow up guys, that all makes sense, happy we can explain the problem now! 🙂 

There's just one final thing I don't quite understand. In the test, the DynamicMethod is created 3 times: one for each context. But as I understand it, we're always creating the dynamic module, with the same name (`"KafkaCachedMessageHeadersHelpers"`) in the _same_ `typeof(DuckType).Module`. Why does that not cause name conflicts (or similar) in the module? Is it because the `typeof(DuckType).Module`s end up being in different contexts too (even though we didn't explicitly use `Assembly.LoadFile`/`AssemblyLoadContext` for that assembly)?

Reply:
I think this is okay because it's not a global name, it's just a name used for stack traces. In other parts of the duck type library we have to be careful and use unique names for proxy types that are created